### PR TITLE
support anon consts in binders

### LINKED ISSUE #72129: Post lazy normalization for consts cleanup (#71973)
While #71973 implements lazy normalization for constants,
there are still some bugs we need to deal with afterwards.

This issue is a collection of things not explicitly mentioned with `FIXME(lazy_normalization_consts)` so we don't forget anything.

### PR DESCRIPTION
**Pull request by lcnr**
closes #72129

The first 2 commits are taken from #79298 which are used in some of the added tests.

This allows us to deal with anonymous constants inside of binders, for example `struct Foo<T>(T) where for<'a> [T; { let _: &'a (); 3 }]: Sized;`.

What's the problem and how am I trying to solve it?

Looking at the above example, `'a` is bound region. As we typeck the anonymous constant separately we would pretty much always work inside of the binder and having bound regions is not supported there.

What we instead want is to treat `'a` inside of the anon const as an early bound region. To achieve this we to add some additional generic params to the created `AnonConst`.

So we pretty much want to desugar the above example into.
Note that we add another generic param to `ANON_CONST`.
```rust
const ANON_CONST<T, 'a> = { let _: &'a (); 3 };
struct Foo<T>(T) where for<'a> [T; ANON_CONST::<T, 'a>]: Sized;
```

To achieve this we extend `hir::AnonConst` to remember both it's new synthetic generic parameters as well as the potentially use lifetime arguments originating from a binder. This is now working well enough to warrant a review.

Once the general approach is ironed out I still want to update lifetime resolution to warn for `for<'a> [u8; 3 + 4]: Sized` that `'a` is unused. We also still have to think a bit about perf, though we can probably put the generics into a optional allocation.

r? @matthewjasper

### CODE REVIEW

File: compiler/rustc_trait_selection/src/traits/wf.rs [hunk 1]
```diff
@@ -434,15 +434,19 @@ impl<'a, 'tcx> WfPredicates<'a, 'tcx> {
...
-                            let cause = self.cause(traits::MiscObligation);
-                            self.out.push(traits::Obligation::with_depth(
-                                cause,
-                                self.recursion_depth,
-                                self.param_env,
-                                predicate,
-                            ));
+                            // FIXME(const_generics): Do we actually end up having to deal with
+                            // this bound later or is this unsound?
+                            if !substs.has_escaping_bound_vars() {
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/wf.rs:439]: cc @eddyb 

Can we somehow circumvent this? :thinking:

I don't think we can just replace the bound vars with placeholders here either.

I think that places which actually try to evaluate the constant do still have to satisfy the `ConstEvaluatable` predicate but it does allow for 
```rust
// check-pass
#![feature(const_generics)]
#![allow(incomplete_features)]

trait Baz<const N: usize> {}

fn test<T>() {
    // FIXME(const_generics): This should error.
    let _a: Box<dyn for<'a> Baz<{
        let _: &'a ();
        std::mem::size_of::<T>()
    }>>;
}

fn main() {
    test::<u32>();
}
```
which is less than ideal

**Comment by lcnr** [comment]
cc @varkor @eddyb @nikomatsakis, this might also be interesting to you

File: compiler/rustc_hir/src/hir.rs [hunk 2]
```diff
@@ -1371,7 +1373,7 @@ pub struct Expr<'hir> {
 
 // `Expr` is used a lot. Make sure it doesn't unintentionally get bigger.
 #[cfg(target_arch = "x86_64")]
-rustc_data_structures::static_assert_size!(Expr<'static>, 72);
+rustc_data_structures::static_assert_size!(Expr<'static>, 144);
```
Comment (cjgillot) [inline compiler/rustc_hir/src/hir.rs:1376]: Would replacing `AnonConst` by `&'hir AnonConst<'hir>` help with the size?

Comment (lcnr) [reply compiler/rustc_hir/src/hir.rs:1376]: yeah, probably :+1: 

I want to wait for a first review before looking into perf though, so I don't waste time on an approach we don't want in the end

**Comment by Dylan-DPC** [comment]
@lcnr any updates on this?

**Comment by lcnr** [comment]
want to talk about this with project-const-generic in the near future before doing much here, going to mark this as blocked on team until then

**Comment by lcnr** [comment]
probably want to wait with this pr until @rust-lang/project-const-generics figures out how to deal with unused params in general as that will influence the way we want to implement this change.

**Pull request closed without merging**